### PR TITLE
HDDS-5517. Support multiple container moves from a source datanode in one balance iteration

### PR DESCRIPTION
**Pull request by JacksonYao287**
## What changes were proposed in this pull request?

Support multiple container moves from a source datanode

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5517

## How was this patch tested?

UT

### CODE REVIEW

**Comment by JacksonYao287** [comment]
@lokeshj1703 @siddhantsangwan  please take a look at this jira as a priority , i think it will speed up the move process very much.
by the way , i think we should not limit `maxSizeEnteringTarget` and `maxSizeLeavingSource` by default, we can specify the value in the command line if needed

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:97]: I think we should not add this map `remainingUsageBeforeOverUtilized`. We can probably use sizeLeavingNode here.

**Review by lokeshj1703 — COMMENTED** [review]
@JacksonYao287 Thanks for working on this!
Currently we iterate over overutilized nodes. Maybe it is better to have a selection criteria for source datanodes as well. Default implementation would be to just return the DN with highest utilisation (also including size being moved).

**Comment by JacksonYao287** [comment]
>Maybe it is better to have a selection criteria for source datanodes as well. Default implementation would be to just return the DN with highest utilisation (also including size being moved).

thanks @lokeshj1703 for the review. the suggestion looks good , i will do this in a new commit

**Comment by JacksonYao287** [comment]
@lokeshj1703 @siddhantsangwan @ChenSammi PTAL, thanks!

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:507]: This block of code is analogous to what `canSizeEnterTarget` does for target nodes. Can we have a similar method called `canSizeLeaveSource` for source nodes?
Also, the `calculateUtilization(long plusSize)` method adds the specified plus size to used space. Here, I think we want to subtract `totalLeavingSize` from used space, instead.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:700]:
```suggestion
               .calculateUtilization(sizeEnteringAfterMove), upperLimit) <= 0;
```

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/SourceDataNodeSelectionCriteria.java
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/SourceDataNodeSelectionCriteria.java:59]: My understanding of this method is unclear, so I might be wrong. I think we want to sort by reducing the used space (subtracting `sizeLeavingNode`) and then calculating utilization.

**Review by siddhantsangwan — COMMENTED** [review]
@JacksonYao287 The changes mostly look good. How about we make the source datanodes selection criteria methods a part of the existing `ContainerBalancerSelectionCriteria` class instead of creating a new one?

**Comment by siddhantsangwan** [comment]
Also, please note the following concern that was raised [here](https://github.com/apache/ozone/pull/2230#discussion_r630705685) for an earlier version of balancer. 

> Question:
> 
> say we have a 10 DN cluster, the usage of all of them is 95%, then one empty DN is added to rebalance the cluster. Given the threshold is 10%, it seems the balancer will not work in this case, since that 10 DN will not achieve upperLimit.
> 
> Have we consider corner case like this ?

By removing `withinThresholdUtilizedNodes`, we rely on the user setting a suitable threshold to make balancing work in such cases.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:266]: sorry for this, i did not meant to remove this comment, i will add it back

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:507]: thanks for pointing out this,  will fix it!

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:700]: the return type of `Double.compare` is int , so `< 1`  is equal to `<= 0` here. but `<=0` seems more clear, will fix this, thanks.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/SourceDataNodeSelectionCriteria.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/SourceDataNodeSelectionCriteria.java:59]: >I think we want to sort by reducing the used space (subtracting sizeLeavingNode) and then calculating utilization.

yes , that is correct. `getNextCandidateSourceDataNode` always try to return a source data node with the highest usage. thanks very much for pointing out this mistake!

**Comment by JacksonYao287** [comment]
thanks very much @siddhantsangwan for the review, i have updated the patch , please take a look!
>By removing withinThresholdUtilizedNodes, we rely on the user setting a suitable threshold to make balancing work in such cases.

please take a look at this explanation which i left at the slack channel `ozone-balancer-discussion`
https://the-asf.slack.com/archives/C02K6DA2UNP/p1636440682032900

**Comment by JacksonYao287** [comment]
> How about we make the source datanodes selection criteria methods a part of the existing `ContainerBalancerSelectionCriteria` class instead of creating a new one?

looks good, so that we have only one criteria for all the selections. what do you think @lokeshj1703 ?

**Review by lokeshj1703 — COMMENTED** [review]
@JacksonYao287 I think there are three different changes being made in this PR. It is better to separate these into different PRs. 
1. I was thinking we should have sth similar to FindTargetStrategy for source like FindSourceStrategy.
2. Let's have a separate PR for withinThreshold nodes removal. I think there was a use case where accounting them was important where there are 5 over utilised and 5 within threshold nodes. cc @siddhantsangwan I think user shouldn't have to adjust threshold in this case.  
3. Also by default we shouldn't allow all size of data to be moved from source/target. It is important to limit otherwise it is possible for all balancing to happen in 2-4 nodes.

**Comment by siddhantsangwan** [comment]
I think we should keep within threshold nodes. Removing them implies the user has to do a lot of work in first analyzing the cluster and then calculating a suitable threshold. The user can understandably expect to introduce one new node to the cluster and have balancer balance it with the default threshold. 

I don't see any down side in the current logic related to within threshold nodes. There's a bug that was pointed out by 
@JacksonYao287:

> in some case ,  one container may be moved from a withinThresholdUtilized  node to another withinThresholdUtilized   node

An easy fix for this is to remove `withinThresholdNodes` from `potentialTargets` when matching within threshold nodes with under utilized nodes.

**Comment by JacksonYao287** [comment]
thanks @lokeshj1703  and @siddhantsangwan for the review!
>1. I was thinking we should have sth similar to FindTargetStrategy for source like FindSourceStrategy.

we can add an interface of FindSourceStrategy and add a default implementation for it. this can be done by refactoring `SourceDataNodeSelectionCriteria`.

>2. Let's have a separate PR for withinThreshold nodes removal. I think there was a use case where accounting them was important where there are 5 over utilized and 5 within threshold nodes. 

sure, let us create another to solve this, and fix related bug. in this patch , we focus supporting multiple container moves from a source datanode.  i will add  withinThreshold nodes back in a new commit

>3. Also by default we shouldn't allow all size of data to be moved from source/target. It is important to limit otherwise it is possible for all balancing to happen in 2-4 nodes.

in this patch , before matching a target with a source , we will sort all the target datanodes In ascending order and all the source datanodes in descending  order by usage rate considering `SizeEnteringnode` and `SizeLeavingNode`, so that we always try to match a source datanode with the biggest usage to the target datanode with a smallest usage.  it is Almost impossible for all balancing to happen in only 2-4 nodes， unless those 2-4 nodes has a much bigger or smaller usage than others. i think if that case happens, it makes sense to move only among these Very unbalanced datanodes

**Comment by JacksonYao287** [comment]
@lokeshj1703 @siddhantsangwan i have refactored the patch according to the comments, please take a look. 

>in some case , one container may be moved from a withinThresholdUtilized node to another withinThresholdUtilized node

i will create a separate patch to add withinThreshold nodes back into candidate target and source datanodes, and fix the potential bug in the newly refactored code.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:354]: Can we avoid calling reinitialize in every iteration?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:494]: Can we move this to selection criteria#getCandidateContainers? Maybe selection criteria can have reference to find source strategy.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java:92]: We can probably use a priority queue here or a TreeSet. TreeSet might be better.

**Review by lokeshj1703 — COMMENTED** [review]
@JacksonYao287 The changes look really good. I have a few comments inline.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (lokeshj1703) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:354]: Is this required for recalculating limits and other metadata in a later PR?

**Comment by lokeshj1703** [comment]
Regarding the max limits - Lets consider a use case. Dn1-3 has 100TB, dn4-6 50TB and dn7-10 are new nodes. In this case all balancing would happen with dn1-3 only. Max limits helps in limiting the amount of data replicated from/to dn which I think is important otherwise a dn could be overburdened with too many replications. In the above case if we add 50 new nodes, still all balancing would happen with dn1-3.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:494]: looks good , will do this, thanks!

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:354]: > Is this required for recalculating limits and other metadata in a later PR?

yea, it is. so think we can keep it here in this PR, and i will deal with it in the later coming PR

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java:92]: TreeSet is implemented by a binary search tree. when searching or inserting any element , it will take O(logN) of time.
priority queue is implemented with heap sort, which is very quick to get the top 1 element after initialized, and I have considered using it。
here, we always try to get the element with the biggest usage(top 1). so i think it is better off using priority queue here.

**Comment by JacksonYao287** [comment]
thanks @lokeshj1703 very much for the review, i have updated the patch, please take a look!
>Max limits helps in limiting the amount of data replicated from/to dn which I think is important otherwise a dn could be overburdened with too many replications. 

i do get your point，I have also considered this。but current Max limit is just the same as a default container size(5G), and i think it is too small to support moving more than one container. It depends and is debatable what value to set, so i do not specified it for now. what about 500G by default, or do you have any suggestion? we can set this in the later PR.

**Comment by siddhantsangwan** [comment]
@JacksonYao287 I'll soon start working on a jira to make the default configurations smarter. Let's discuss these values and reset them in that jira?

**Comment by JacksonYao287** [comment]
> I'll soon start working on a jira to make the default configurations smarter. Let's discuss these values and reset them in that jira?

thanks @siddhantsangwan , sounds good,  let`s determine it in that jira.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java:94]: Let's do a priority queue implementation here as well.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java:77]: Let's revert the configuration changes here and do it in a separate PR.

**Review by lokeshj1703 — COMMENTED** [review]
@JacksonYao287 Thanks for updating the PR! I have few minor comments inline.

**Comment by lokeshj1703** [comment]
Regarding the max limits, I don't think we can support very large limits. 500 G means 500GB data can move from/to datanode in one iteration. I am not sure how much replication dn can support per minute. Maybe it should be determined first.
Further we also have move timeouts, I think move would definitely time out with this much payload.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java:77]: sure , will do this!

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java:94]: i think we should use TreeSet here.
in `FindSourceGreedy`, we always try to get the top one source, and if no target and be matched with it , we will remove it.
but here, we never remove any target , and just try to find a target with Middle-order traversal。but we can not do  traversal on priorityqueue.

**Comment by JacksonYao287** [comment]
thanks @lokeshj1703 for the review, i have updated this patch , please take a look.

> Regarding the max limits, I don't think we can support very large limits. 500 G means 500GB data can move from/to datanode in one iteration. I am not sure how much replication dn can support per minute. Maybe it should be determined first. Further we also have move timeouts, I think move would definitely time out with this much payload.

we can discuss this in @siddhantsangwan `s jira, which will make the default configurations smarter

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java:70]: Sorry for not bringing this up earlier! But we will need to handle the case in the comparator where utilisation is same for two nodes.

Comment (lokeshj1703) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java:70]: +1 o.w.

Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java:70]: good point , thanks for pointing out this !

Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindTargetGreedy.java:70]: good point , thanks for pointing out this

**Comment by JacksonYao287** [comment]
@lokeshj1703 thanks for pointing out the mistake , i have updated this patch , please take a look

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java:56]: Can we do datanode details UUID comparison instead?

Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java:56]: sure, will do this.

**Review by lokeshj1703 — APPROVED** [review]
Pending clean CI.

**Comment by JacksonYao287** [comment]
@lokeshj1703  @siddhantsangwan thank you for the review! CI failure seems not caused by this patch. 
i have tested this patch in my Kubernetes cluster, and it works as expected。the process of balancing is greatly accelerated。

**Comment by lokeshj1703** [comment]
@JacksonYao287 We will have to get a clean CI before merge. That is the process we follow for Ozone. Maybe you can try rebasing the PR on current master.

**Comment by JacksonYao287** [comment]
@lokeshj1703 sure, thanks , i have merged current  master branch into this patch. let`s wait for a clean CI

**Comment by lokeshj1703** [comment]
@JacksonYao287 Thanks for the contribution! @siddhantsangwan  Thanks for review! I have committed the PR to master branch.

**Comment by JacksonYao287** [comment]
thanks @lokeshj1703 and @siddhantsangwan for the review!

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java:91]: Hey @JacksonYao287 Can you help me understand what's happening in this method? I don't think the usage info for a node will get updated during an iteration, since DU/DF don't run during an iteration.

Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java:91]: let me explain this. 
when selecting a source datanode, we always want to the select the one which has a largest storage usage. here, i use `PriorityQueue`, which is fast to get the top one. when calling `getNextCandidateSourceDataNode`, `PriorityQueue#poll` is called , which will get and remove the top source data node for this PriorityQueue.

actually, now we supporting move multiple containers from one datanode , so a data node can be selected as source for multiple times in one iteration. 

here are two reasons to call `potentialSources.add(nodeManager.getUsageInfo(dui));` 
1 add the data node back  to the PriorityQueue again, so it can be selected as a source again.
2 when we update `sizeLeavingNode`,  the usage of this data node will be considered changed (the reported usage - sizeLeaving), so we need to sort all the candidate source datanodes according to the latest usage and get the top one. when adding the data node back to the  PriorityQueue, PriorityQueue will sort all the datanode again(it will use heap sort , so very fast), so we can get the next top one.

i am not sure is it clear to you now?

**Pull request merged**
